### PR TITLE
[channel loss]support packing & padding free

### PR DESCRIPTION
**Pull request by kevssim**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information

Write the detail information belongs to this PR.

## Experiment results

Paste your experiment result here(if needed).

### CODE REVIEW

File: swift/plugin/loss.py [hunk 1]
```diff
@@ -401,17 +406,34 @@ def channel_loss_func(outputs, labels, num_items_in_batch=None, sample_channels=
...
     flat_logits = shift_logits.view(-1, shift_logits.size(-1))
     flat_labels = shift_labels.view(-1)
-    token_loss = loss_fct(flat_logits, flat_labels).view_as(shift_labels)  # [batch_size, seq_len-1]
-
-    mask = shift_labels != -100
+    token_loss = loss_fct(flat_logits, flat_labels)
+    mask = flat_labels != -100
+
+    if position_ids is not None and logits.size(0) == 1:
+        pos = position_ids[..., :-1].view(-1).tolist()
```
Comment (Jintao-Huang) [inline swift/plugin/loss.py]: 这个可以向量化嘛

参考这里：https://github.com/modelscope/ms-swift/blob/3f4b4f18a8e85ceb13b885c43a8a28b7ac22216b/swift/trainers/mixin.py#L542

方便加个 trainer.template.padding_free的检测不

Comment (kevssim) [reply swift/plugin/loss.py:413]: trainer.template.padding_free的检测是什么意思呢

Comment (kevssim) [reply swift/plugin/loss.py]: 向量化：已修改。
检测：因为packing和padding_free都需要走这部分处理逻辑，所以改为用trainer.template._packing判断，ok不
